hack/local-test.sh

```diff
@@ -0,0 +1,194 @@
+      return 1
+    fi
+
+    for task_version in "${task_versions[@]}"
+    do
+      modified_task_file="/tmp/${task_file##*/}-modified"
+      yq ".metadata.labels.\"app.kubernetes.io/version\" |= \"${task_version}\"" "$task_file" >"$modified_task_file"
+      annotations=()
+      migration_file="hack/task/${task_name}/migrations/${task_version}.sh"
+      if [[ -e "$migration_file" ]]; then
+          annotations+=(--annotate "${MIGRATION_ANNOTATION}=true")
+      fi
+	
+      sha1_hash=$(generate_sha1_hash "$task_version")
+      image_tag="${task_version%.*}-${sha1_hash}"
+      image_with_tag="$image:$image_tag"
+      echo "info: push task bundle $image_with_tag"
+      output=$(tkn_bundle_push "${image_with_tag}" -f "$modified_task_file" "${annotations[@]}")
+      echo "$output"
+      bundle_ref=$(echo "$output" | grep "sha256:[0-9a-f]\+" | cut -d' ' -f5)
+
+      if [[ -e "$migration_file" ]]; then
+        echo "info: attach migration file ${migration_file} to task bundle ${bundle_ref}"
+        oras attach --registry-config "$REGISTRY_AUTH_JSON" \
+          --annotation "${MIGRATION_ANNOTATION}=true" \
+          --artifact-type "$MIGRATION_ARTIFACT_TYPE" \
+          --distribution-spec v1.1-referrers-tag \
+          "$bundle_ref" \
+          "$migration_file"
+      fi
+
+
+      if [[ "$task_name" == "clone" ]]; then
+        clone_upgrades_data+=("$image $image_tag ${bundle_ref##*@}")
+      elif [[ "$task_name" == "tests" ]]; then
+        tests_upgrades_data+=("$image $image_tag ${bundle_ref##*@}")
+      fi
+    done
+}
+
+
+remove_task_bundles() {
+  local image
+  local -a tags
+  for task_name in "${TASK_NAMES[@]}"; do
+    image="${IMAGE_PREFIX}/task-${task_name}"
+    echo "info: listing tags from image repository ${image}"
+    mapfile -t tags < <(skopeo list-tags "docker://${image}" | jq -r '.Tags[]')
+    for tag_name in "${tags[@]}"; do
+      echo "info: remove image tag ${image}:${tag_name}"
+      skopeo delete "docker://${image}:${tag_name}"
+    done
+  done
+}
+
+
+make_renovate_upgrades() {
+    local upgrade_from upgrade_to
+    local n
+    local dep_name current_value current_digest new_value new_digest
+
+    local -r data_file="/tmp/pmt-upgrades-data.txt"
+
+    echo -n "[" | tee "$data_file"
+
+    upgrade_from=${clone_upgrades_data[0]}
+    dep_name=$(cut -d' ' -f1 <<<"$upgrade_from")
+    current_value=$(cut -d' ' -f2 <<<"$upgrade_from")
+    current_digest=$(cut -d' ' -f3 <<<"$upgrade_from")
+    n=${#clone_upgrades_data[@]}
+    n=$((n-1))
+    upgrade_to=${clone_upgrades_data[$n]}
+    dep_name=$(cut -d' ' -f1 <<<"$upgrade_to")
+    new_value=$(cut -d' ' -f2 <<<"$upgrade_to")
+    new_digest=$(cut -d' ' -f3 <<<"$upgrade_to")
+
+      echo "{
+  \"depName\": \"${dep_name}\",
+  \"currentValue\": \"${current_value}\",
+  \"currentDigest\": \"${current_digest}\",
+  \"newValue\": \"${new_value}\",
+  \"newDigest\": \"${new_digest}\",
+  \"packageFile\": \".tekton/ci-pipeline.yaml\",
```

Comment:
This has been changed to another file, that exists in a local directory git ignores.

Reply:
Hmm, so how should I run the tool and have it succeed? I can probably figure it out, but it would be good to commit the target file or at least have instructions for how to create it in the README

Reply:
And while testing it, it seemed like the changes weren't getting applied :thinking: 

The file did get re-formatted, but the migrations didn't add the annotation

```diff
diff --git a/local-tests/ci-pipeline.yaml b/local-tests/ci-pipeline.yaml
index 814523b..bbc704e 100644
--- a/local-tests/ci-pipeline.yaml
+++ b/local-tests/ci-pipeline.yaml
@@ -8,7 +8,7 @@ metadata:
     build.appstudio.redhat.com/target_branch: '{{target_branch}}'
     pipelinesascode.tekton.dev/max-keep-runs: "3"
     pipelinesascode.tekton.dev/on-cel-expression: event == "pull_request" && target_branch == "redhat-latest"
-  creationTimestamp: null
+  creationTimestamp:
   labels:
     appstudio.openshift.io/application: rh-syft
     appstudio.openshift.io/component: rh-syft
@@ -160,13 +160,13 @@ spec:
         value: $(params.git-url)
       - name: revision
         value: $(params.revision)
-      # test fixtures include symlinks that point outside the repo
+          # test fixtures include symlinks that point outside the repo
       - name: enableSymlinkCheck
         value: 'false'
-      # tags are crucial for determining versions
+          # tags are crucial for determining versions
       - name: fetchTags
         value: 'true'
-      # tags don't work with a shallow clone, do a full clone
+          # tags don't work with a shallow clone, do a full clone
       - name: depth
         value: '0'
       runAfter:
@@ -399,7 +399,7 @@ spec:
   - name: workspace
     volumeClaimTemplate:
       metadata:
-        creationTimestamp: null
+        creationTimestamp:
       spec:
         accessModes:
         - ReadWriteOnce

``

Reply:
I think it depends on what commands written inside the migration file. Do you have `.metadata.annotations.changes` annotation in the pipeline?

Reply:
I tried it with 

```yaml
kind: Pipeline
metadata:
  annotations:
    changes: ""
```

^ works as expected

```yaml
kind: PipelineRun
spec:
  pipelineSpec:
    metadata:
      annotations:
        changes: ""
```
^ doesn't work :confused:

Reply:
Oh nevermind, I figured it out. The migrations don't do anything because they edit the `.metadata`. But for a `PipelineRun`, the only way the changes will be preserved is if they edit the `.spec`

Reply:
Maybe just change the example migrations to edit something under `.spec`?

Reply:
Updated. PTAL.

Reply:
Very nice :+1: